engine/gameWorker.js

```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
 			});
 		}
 
-		function begin(config) {
+		function sendIncomplete() {
+			//Call us back, we want to continue on
+			const now = Date.now();
+			const state = game.getState();
+			self.postMessage({
+				action: 'STEP_INCOMPLETE',
+				state: Object.assign({}, state, {
+					realWorldTime: now - time0,
+				}),
+			});
+		}
+
+		function begin(config, checkbackTime) {
 			if(game) {
 				throw new Error('Cannot re-use game worker');
 			}
 			time0 = Date.now();
 			game = new GameManager(new Random(config.seed), config);
+			//I consider this a hack, but it's hard to figure out a better way
+			if (config.hasOwnProperty('startFrame') && config.startFrame > 0) {
+				let limit = 0;
+				if (checkbackTime) {
+					let prevLim = Math.floor(Date.now()/checkbackTime)*checkbackTime;
+					limit = prevLim + checkbackTime;
+				}
+				try {
+					for(let i = 0; (i < config.startFrame) && !game.isOver(); ++ i) {
+						game.step(null);
+						/* jshint maxdepth:6 */
+						if(limit && Date.now() >= limit) {
+							let prevLim = Math.floor(Date.now()/checkbackTime)*checkbackTime;
+							limit = prevLim + checkbackTime;
+							sendIncomplete();
+						}
+					}
+				} catch(e) {
+					if(e === 'PAUSE') {
+						sendState(true);
+					} else {
+						throw e;
+					}
+				}
+			}
 			sendState();
 		}
 
-		function step({maxTime, steps, type}) {
-			const limit = maxTime ? (Date.now() + maxTime) : 0;
+		function step({checkbackTime, steps, type}) {
+			let limit = 0;
+			if (checkbackTime) {
+				limit = Math.floor(Date.now()/checkbackTime)*checkbackTime + checkbackTime;
```

Comment:
Timing against intervals naively implies accepting whatever latency between timer expiration and response there is, letting clock drift happen. Yes, this can be accounted for with some bookkeeping, but timing against a deadline handles this issue naturally.

Reply:
I think the difference in viewpoints here is because you're considering it from the perspective of a long-running task which we get updates from at regular intervals but don't need to control, whereas I'm thinking of a user who pressed the "turbo speed" button and wants the simulation to run as quickly as possible, but still leave them in control. These feel different enough in behaviour that it might be worth having 2 separate controls; one for regular update intervals (used when skipping and when running games in the background during a tournament), and the other for running as many frames as possible within a given amount of time (turbo speed).